src/Middleware/DetectLocaleMiddleware.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace TractorCow\Fluent\Middleware;
+
+use SilverStripe\Control\Cookie;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\Middleware\HTTPMiddleware;
+use SilverStripe\Core\Config\Configurable;
+use TractorCow\Fluent\Extension\FluentDirectorExtension;
+use TractorCow\Fluent\Model\Locale;
+use TractorCow\Fluent\State\FluentState;
+use TractorCow\Fluent\State\LocaleDetector;
+
+/**
+ * DetectLocaleMiddleware will detect if a locale has been requested (or is default) and is not the current
+ * locale, and will redirect the user to that locale if needed.
+ */
+class DetectLocaleMiddleware implements HTTPMiddleware
+{
+    use Configurable;
+
+    /**
+     * IDs to persist the locale in cookies / session in the front end, CMS, etc
+     *
+     * @config
+     * @var string[]
+     */
+     #
+    private static $persist_ids = [
+        'frontend' => 'FluentLocale',
+        'cms' => 'FluentLocale_CMS',
+    ];
+
+    /**
+     * The expiry time in days for a locale persistence cookie
+     *
+     * @config
+     * @var int
+     */
+    private static $persist_cookie_expiry = 90;
+
+    /**
+     * Whether cookies have already been set during {@link setPersistLocale()}
+     *
+     * @var bool
+     */
+    protected $cookiesPersisted = false;
+
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        $localeDetector = LocaleDetector::singleton()->setRequest($request);
```

Comment:
I suggest we make these services stateless; Instead of setRequest, we pass this in via each method that requires it.